openpgp/packet/signature.go

```diff
@@ -539,7 +547,9 @@ func (sig *Signature) signPrepareHash(h hash.Hash) (digest []byte, err error) {
 // On success, the signature is stored in sig. Call Serialize to write it out.
 // If config is nil, sensible defaults will be used.
 func (sig *Signature) Sign(h hash.Hash, priv *PrivateKey, config *Config) (err error) {
-	if priv.PrivateKey == nil {
+	signer, hashIsSigner := h.(Signer)
+
+	if !hashIsSigner && priv.PrivateKey == nil {
```

Comment:
I added a test in b5bb787789c47b9d53f45610eeb3cdcc8195afb6, checking that passing a non-Signer `h` and a `nil` `priv` will trigger this error.